### PR TITLE
Refactor move session client

### PR DESCRIPTION
**Pull request by adshmh**
TLDR: This PR moves Session-related functionalities out of `sdk.go` and into a top-level `session.go`. It is placed in draft mode because it includes changes from #10 which need to be merged before reviewing this PR.

Main changes made:
- Remove HeightClient and make the latest height an input argument for several functions, both to simplify the code by getting rid of an interface and to allow more control over fetching the latest block height from a node.
- Make SessionClient a concrete struct that can be used to get a session.
- Introduce a PoktrollSessionFetcher interface to clarify the interactions of the SessionClient with poktroll nodes and allow customization of the said interaction: most users will not need this and can simply use the NewSessionClient function.
- Refactor GetSessionSupplierEndpoints:
     - Accept a session instead of calling a SessionClient, to make interactions with Poktroll nodes more efficient and allow more control over them if needed.
     - Make it a utility function as it doesn't need to make direct calls to Poktroll nodes anymroe.
     - Move it to the new session.go file.
     - Make it use Poktroll's exported types instead of a custom sdk-defined type.
- Remove the now redundant SessionClient interface from sdk.go
- Update the SendRelay function to make the input arguments more granular: this allowed scrapping the custom types defined under types/types.go as the code now uses the Poktroll's exported types.
- Delete the now obsolete types/types.go: using Poktroll's exported types (custom, SDK-specific types are not needed at this point)

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+// This can help enforce type safety by requiring explict type casting of a string before it can be used as a SupplierAddress.
+//
+// ServiceEndpoints returns the supplier endpoints assigned to a session for the given service id.
+//
+// The returned value is a map of supplierId to the list of endpoints.
+func ServiceEndpoints(
+	session *sessiontypes.Session,
+	serviceId string,
+) map[string][]*sharedtypes.SupplierEndpoint {
+	supplierEndpoints := make(map[string][]*sharedtypes.SupplierEndpoint)
```
Comment (red-0ne) [inline session.go:62]: I'd rather keep the `SessionSuppliers` struct and use it to return the available suppliers data to be used by `SendRelay`

* It will ensure that a matching `SupplierAddress/SupplierUrl` is provided to `SendRelay` instead of potentially dissociated ones.
* It also returns all the supplier urls while retaining the corresponding `RPCType` info.
* A matching `SessionHeader` is always accompanying the `Supplier` provided.

File: sdk/sdk.go [hunk 2]
```diff
@@ -151,22 +94,25 @@ func (sdk *ShannonSDK) GetApplicationsDelegatingToGateway(
 // response after verifying the supplier's signature.
 func (sdk *ShannonSDK) SendRelay(
 	ctx context.Context,
-	sessionSupplierEndpoint *types.SingleSupplierEndpoint,
+	header *sessiontypes.SessionHeader,
```
Comment (red-0ne) [inline sdk/sdk.go:97]: I prefer providing `types.SingleSupplierEndpoint` which could ensure matching `SessionHeader`, `SupplierAddress` and `EndpointUrl` instead of giving unnecessary responsibility to the consumer to collect/match and provide these independently.

File: session.go (hunk 1 above)
Comment (adshmh) [reply session.go:103]: Great point, thank you.
Since endpoint selection is a feature that is of interest to Gateways, and is always done in the context of a session, a better approach, compared to the first version of the PR, is to use a wrapper struct. I have updated the PR with a first version of this approach. PTAL.

File: sdk/sdk.go (hunk 2 above)
Comment (adshmh) [reply sdk/sdk.go]: Thank you for the review. Updated, PTAL.

File: session.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		supplierEndpoints[supplier.Address] = endpoints
+	}
+
+	return supplierEndpoints, nil
+}
+
+// SelectEndpoint adds the specifed supplier+endpoint combination to the list of selected endpoints of a service, e.g. for sending relays.
+// This method is not safe to use concurrently by multiple goroutines.
+func (f *FilteredSession) AddEndpointToSelection(serviceId string, supplierAddress string, endpoint *sharedtypes.SupplierEndpoint) error {
```
Comment (red-0ne) [inline session.go:130]: I see a few problems with this method.
1. It adds unneeded statefulness to the `FilteredSession`. I see that passing a filter function would be more elegant.
2. It can accumulate selected `SupplierEndpoint`s but I see no method to retrieve them all, `SelectedEndpoint()` only returns the first.
3. It implies that the caller has already selected the endpoint to add and has the responsibility to pass in the matching `supplierAddress` (I really prefer a custom struct that contains wraps the endpoint with its corresponding supplier if we do not opt of passing a filter function).

File: session.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+// PoktNodeSessionFetcher is used by the SessionClient to fetch sessions using poktroll request/response types.
+//
+// Most users can rely on the default implementation provided by NewPoktNodeSessionFetcher function.
+// A custom implementation of this interface can be used to gain more granular control over the interactions
+// of the SessionClient with the POKT full node.
+type PoktNodeSessionFetcher interface {
+	GetSession(
+		context.Context,
+		*sessiontypes.QueryGetSessionRequest,
+		...grpcoptions.CallOption,
```
[the remark this answers is not in the document]
Comment (adshmh) [reply session.go:67]: I agree: this is only here so we can directly use the Session module's query client.

File: session.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+// TODO_DISCUSS: Consider using a custom type, defined as a string, as supplier address.
+// This can help enforce type safety by requiring explict type casting of a string before it can be used as a SupplierAddress.
+//
+// ServiceEndpoints returns the supplier endpoints assigned to a session for the given service id.
+//
+// The returned value is a map of supplierId to the list of endpoints.
+func (f *FilteredSession) ServiceEndpoints(
+	serviceId string,
```
[the remark this answers is not in the document]
Comment (adshmh) [reply session.go:97]: Thank you for the review: based on my reading of the `Session` struct, it can contain multiple suppliers, and each supplier could contain multiple services, hence the need for specifying a `serviceId` (which I agree looks redundant, but may actually be needed due to how the Supplier struct works). Is this accurate?

[the remark this answers is not in the document]
Comment (red-0ne) [reply session.go:97]: We are sure that the suppliers contained in the `Session` provide the service that the `Session` is about. It's just that the on-chain part is not removing the unrelated services from those Suppliers.

Which means that all `Supplier`s in the `Session` have at least 1 endpoint related to the `serviceId`.

I think this method is a good place to remove the unrelated endpoints and return only the relevant ones.

**Review by red-0ne — APPROVED** [review]
Love the shape it's taking :fire: 

LGTM!
